tests/tests/storage_read_replay.rs

```diff
@@ -0,0 +1,178 @@
+        op::call(0x10, RegId::ZERO, RegId::ZERO, RegId::CGAS),
+        op::log(0x10, RegId::ZERO, RegId::ZERO, RegId::ZERO),
+        op::ret(RegId::ONE),
+    ];
+
+    let mut script_data = contract_id.to_vec();
+    script_data.extend(0u64.to_be_bytes());
+    script_data.extend(0u64.to_be_bytes());
+
+    let tx = TransactionBuilder::script(script.into_iter().collect(), script_data)
+        .script_gas_limit(gas_limit)
+        .maturity(maturity)
+        .add_unsigned_coin_input(
+            SecretKey::random(rng),
+            rng.gen(),
+            u32::MAX as u64,
+            Default::default(),
+            Default::default(),
+        )
+        .add_input(Input::contract(
+            rng.gen(),
+            rng.gen(),
+            rng.gen(),
+            Default::default(),
+            contract_id,
+        ))
+        .add_output(Output::contract(1, Default::default(), Default::default()))
+        .finalize_as_transaction();
+
+    let mut status_stream = client.submit_and_await_status(&tx).await.unwrap();
+    let intermediate_status = status_stream.next().await.unwrap().unwrap();
+    assert!(matches!(
+        intermediate_status,
+        TransactionStatus::Submitted { .. }
+    ));
+    let final_status = status_stream.next().await.unwrap().unwrap();
+    let TransactionStatus::Success { block_height, .. } = final_status else {
+        panic!("Tx wasn't included in a block: {:?}", final_status);
+    };
+    block_height
+}
+
+/// Create a counter contract.
+/// Increment it multiple times, and make sure the replay gives correct storage state every time.
+#[tokio::test(flavor = "multi_thread")]
+async fn test_storage_read_replay_returns_counter_state() {
+    let mut rng = rand::rngs::StdRng::seed_from_u64(0xBAADF00D);
+
+    let mut node_config = Config::local_node();
+    node_config.debug = true;
+    let srv = FuelService::new_node(node_config.clone()).await.unwrap();
+    let client = FuelClient::from(srv.bound_address);
+
+    let (contract_id, block_height) = make_counter_contract(&client, &mut rng).await;
+
+    let _replay = client
+        .storage_read_replay(&block_height)
+        .await
+        .expect("Failed to replay storage read");
+
+    let mut storage_slot_key = contract_id.to_vec();
+    storage_slot_key.extend(Bytes32::zeroed().to_vec());
+
+    for i in 0..10u64 {
+        let block_height = increment_counter(&client, &mut rng, contract_id).await;
+
+        let replay = client
+            .storage_read_replay(&block_height)
+            .await
+            .expect("Failed to replay storage read");
+
+        let value = replay
+            .iter()
+            .find(|item| item.column == "ContractsState" && item.key == storage_slot_key)
+            .expect("No storage read found")
+            .value
+            .clone();
+
+        let mut expected_value = [0; 32];
+        expected_value[..8].copy_from_slice(&i.to_be_bytes());
+
+        assert!(value == Some(expected_value.to_vec()));
```

Comment:
Could create a small helper:
```rs
fn get_counter_from_storage_bytes(bytes: &[u8]) -> u64 {
    let counter_bytes = bytes[..8];
    u64::from_be_bytes(counter_bytes)
}
```
and then
```rs
let value = (...).unwrap();
let actual = get_counter_from_storage_bytes(&value);
let expected = i;

assert_eq!(expected, actual);
```

Reply:
fc097bf56e